### DISCUSSION TITLE
Load pages eagerly on change in `next dev`

### DISCUSSION THREAD

**Discussion by connor4312** [vercel/next.js#23916] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

From: https://github.com/microsoft/vscode-js-debug/issues/730#issuecomment-690622681

> Next.js still does some weird things. Well, things that make life difficult for [the VS Code JS debugger]. Namely, when they recompile, the new source is not actually loaded until you try to load the page, so during this time the file on your disk and the script in the runtime are different. During this period, breakpoints are likely to appear at incorrect locations, since we'll still be loading the old sourcemap.
>
>Also, because of this, we do not know that the page's source updated until it's about to run, so on the first refresh after a change it's likely breakpoints will get missed. Chrome has an option we can use to ask it to pause before executing a script with a sourcemap, but Node.js has no such option.

## Describe the solution you'd like

The cleanest solution to this would be to eagerly load pages on recompile if running in debug mode. Debugging can be easily detecting by checking that [`inspector.url() !== undefined`](https://nodejs.org/api/inspector.html#inspector_inspector_url)

## Describe alternatives you've considered

- One solution that may come to mind is for the debugger to watch for file changes to the sourcemaps. However this only solves part of the problem -- we can detect a change in the sourcemap, but it also needs to be loaded in the runtime, otherwise breakpoints we try to set will still appear at the wrong places. (To set a breakpoint, we convert the "ui" location to the runtime location using sourcemaps, set it in V8, then V8 tells us the location it resolved at, which is mapped back to a "ui" location. If it's set against the old script, the resolution in V8 may fail or resolve to an incorrect location, which will cause the breakpoint to appear at the wrong place in the UI.)

- In the VS Code debugger, we after parsing the sourcemap from the original script the first time, we could set a url (path) breakpoint at the first line of the compiled script so that the runtime will pause before Next.js tries to import it the next time.

    We haven't needed this behavior in other scenarios, but it could also solve the problem and isn't _too_ much of a special case as it would work if another framework used a similar hot-reloading scheme as next. It would add a slight runtime cost to set the breakpoint, but would only actually cause a pause of a hot-reload happens.

    This would be a graceful solution to the issue of not hitting a breakpoint on the first reload, but we would still have misplaced breakpoints until a reload happens.

**Comment by cdaringe** [comment]
We also want to _sometimes_ disable eager compilation--specifically for running tests. Without doing so, each time the test runner hits the page, we need to apply some special, conditional timeout logic. To avoid it, weve just been doing production builds and testing those, but it's less desirable
